File: crates/bindings-typescript/src/sdk/node/container.ts
import { request, type ClientRequest, type IncomingMessage } from 'node:http';
import { isIP, type Socket } from 'node:net';
import { performance } from 'node:perf_hooks';
import { inspect } from 'node:util';
import { Identity } from '../../lib/identity';

const REQUEST_MS = 5_000;
const MAX_BODY = 8192 + 256;
const MAX_HEADERS = 4096;
const LOCAL_AUTHORITY = '127.0.0.1:18081';
const CREDENTIAL_PATH = '/v1/credentials';

export type ContainerCredentialErrorCode =
  | 'missing_environment'
  | 'invalid_discovery'
  | 'invalid_target'
  | 'denied'
  | 'unavailable'
  | 'transport'
  | 'timeout'
  | 'aborted'
  | 'invalid_response';

/** Errors contain no discovery values, response bodies or credentials. */
export class ContainerCredentialError extends Error {
  constructor(readonly code: ContainerCredentialErrorCode) {
    super(`Container credential request failed: ${code}`);
    this.name = 'ContainerCredentialError';
  }
}

type Endpoint =
  | { socketPath: string }
  | { hostname: string; port: number; host: string };

/** Discovery values injected by the container platform; these are not secrets. */
export interface ContainerDiscovery {
  databaseIdentity: string;
  serverUri: string;
  credentialBroker: string;
}

/**
 * Explicit discovery and short-lived credentials for a Node.js process.
 * No stored CLI credentials, default server, anonymous Identity, redirects or
 * proxy configuration are used. Construction performs no I/O.
 */
export class Container {
  #identity: string;
  #server: string;
  #endpoint: Endpoint;

  constructor(discovery: ContainerDiscovery) {
    this.#identity = identity(discovery.databaseIdentity, 'invalid_discovery');
    const server = parseUrl(discovery.serverUri);
    if (!['http:', 'https:', 'ws:', 'wss:'].includes(server.protocol)) {
      throw new ContainerCredentialError('invalid_discovery');
    }
    if (!server.hostname) {
      throw new ContainerCredentialError('invalid_discovery');
    }
    this.#server = server.href;
    this.#endpoint = endpoint(discovery.credentialBroker);
  }

  /** Read only the three platform discovery variables. */
  static fromEnvironment(
    environment: Readonly<Record<string, string | undefined>> = process.env
  ): Container {
    const read = (name: string): string => {
      const value = environment[name];
      if (value === undefined) {
        throw new ContainerCredentialError('missing_environment');
      }
      return value;
    };
    return new Container({
      databaseIdentity: read('SPACETIMEDB_DATABASE_IDENTITY'),
      serverUri: read('SPACETIMEDB_SERVER_URI'),
      credentialBroker: read('SPACETIMEDB_CREDENTIAL_BROKER'),
    });
  }

  get databaseIdentity(): Identity {
    return new Identity(this.#identity);
  }

  get serverUri(): string {
    return this.#server;
  }

  /**
   * Request a fresh token for an exact target Identity, defaulting to this
   * container's database. The broker authenticates the process independently
   * of these discovery values. Send the token only to the target's trusted
   * server. This does not reconnect a database connection or replay calls.
   */
  async tokenFor(
    target: Identity = this.databaseIdentity,
    signal?: AbortSignal
  ): Promise<ContainerToken> {
    let targetHex: string;
    try {
      targetHex = identity(target.toHexString(), 'invalid_target');
    } catch {
      throw new ContainerCredentialError('invalid_target');
    }
    const response = await readToken(this.#endpoint, targetHex, signal);
    return ContainerToken.fromResponse(response, targetHex);
  }

  [inspect.custom](): string {
    return 'Container { discovery: <configured> }';
  }
}

/** An in-memory bearer token. Inspection and JSON serialization are redacted. */
export class ContainerToken {
  #token: string;
  #target: string;
  #expiry: number;
  #deadline: number;

  private constructor(
    token: string,
    target: string,
    expiry: number,
    remaining: number
  ) {
    this.#token = token;
    this.#target = target;
    this.#expiry = expiry;
    this.#deadline = performance.now() + remaining;
  }

  /** @internal */
  static fromResponse(bytes: Buffer, target: string): ContainerToken {
    try {
      const value: unknown = JSON.parse(
        new TextDecoder('utf-8', { fatal: true }).decode(bytes)
      );
      if (!value || typeof value !== 'object' || Array.isArray(value))
        throw new Error();
      const fields = value as Record<string, unknown>;
      const token = fields.token;
      const seconds = fields.expires_unix_seconds;
      if (
        Object.keys(fields).length !== 2 ||
        typeof token !== 'string' ||
        !/^[\x21-\x7e]{1,8192}$/.test(token) ||
        typeof seconds !== 'number' ||
        !Number.isSafeInteger(seconds) ||
        !Number.isSafeInteger(seconds * 1000)
      )
        throw new Error();
      const expiry = seconds * 1000;
      const remaining = expiry - Date.now();
      if (remaining <= 0 || remaining > 30_000) throw new Error();
      return new ContainerToken(token, target, expiry, remaining);
    } catch {
      throw new ContainerCredentialError('invalid_response');
    }
  }

  /** Do not log, persist, or place this value in a URL or environment variable. */
  get value(): string {
    return this.#token;
  }

  get target(): Identity {
    return new Identity(this.#target);
  }

  get expiresAt(): Date {
    return new Date(this.#expiry);
  }

  /** A monotonic cap prevents a backward wall-clock jump from extending it. */
  get remainingLifetimeMs(): number {
    return Math.max(
      0,
      Math.min(this.#deadline - performance.now(), this.#expiry - Date.now())
    );
  }

  toJSON(): object {
    return {
      target: this.#target,
      expiresAt: this.expiresAt,
      token: '<redacted>',
    };
  }

  [inspect.custom](): object {
    return this.toJSON();
  }
}

function identity(value: string, code: ContainerCredentialErrorCode): string {
  if (typeof value !== 'string' || !/^[0-9a-fA-F]{64}$/.test(value)) {
    throw new ContainerCredentialError(code);
  }
  return value.toLowerCase();
}

function parseUrl(value: string): URL {
  try {
    if (
      !value ||
      Buffer.byteLength(value) > 4096 ||
      [...value].some(c => c.charCodeAt(0) <= 32 || c.charCodeAt(0) === 127)
    )
      throw new Error();
    const url = new URL(value);
    if (
      url.username ||
      url.password ||
      value.includes('?') ||
      value.includes('#')
    )
      throw new Error();
    return url;
  } catch {
    throw new ContainerCredentialError('invalid_discovery');
  }
}

function endpoint(value: string): Endpoint {
  const url = parseUrl(value);
  if (url.protocol === 'unix:') {
    const path = value.slice('unix://'.length);
    if (
      !value.startsWith('unix:///') ||
      url.host ||
      !path.startsWith('/') ||
      Buffer.byteLength(path) > 103 ||
      path.includes('%') ||
      path.includes('\\') ||
      path.split('/').some(part => part === '.' || part === '..')
    )
      throw new ContainerCredentialError('invalid_discovery');
    return { socketPath: path };
  }
  // Validate the original authority before URL normalization can turn a DNS
  // spelling or an abbreviated/hexadecimal IPv4 address into loopback.
  const match =
    /^http:\/\/(\[[0-9a-fA-F:]+\]|[0-9.]+)(?::([0-9]+))?\/v1\/credentials$/.exec(
      value
    );
  if (!match) throw new ContainerCredentialError('invalid_discovery');
  const hostname = match[1].replace(/^\[|\]$/g, '');
  const version = isIP(hostname);
  if (
    (version !== 4 || hostname.split('.')[0] !== '127') &&
    (version !== 6 || new URL(`http://[${hostname}]/`).hostname !== '[::1]')
  )
    throw new ContainerCredentialError('invalid_discovery');
  const port = match[2] === undefined ? 80 : Number(match[2]);
  if (!Number.isInteger(port) || port < 1 || port > 65535) {
    throw new ContainerCredentialError('invalid_discovery');
  }
  return { hostname, port, host: url.host };
}

function responseLength(response: IncomingMessage): number {
  if (response.statusCode === 401 || response.statusCode === 403) {
    throw new ContainerCredentialError('denied');
  }
  if (response.statusCode === 503)
    throw new ContainerCredentialError('unavailable');
  if (response.statusCode !== 200)
    throw new ContainerCredentialError('invalid_response');
  const count = (name: string): number =>
    response.rawHeaders.filter(
      (v, i) => i % 2 === 0 && v.toLowerCase() === name
    ).length;
  const headers = response.headers;
  if (
    count('content-type') !== 1 ||
    headers['content-type'] !== 'application/json' ||
    count('cache-control') !== 1 ||
    headers['cache-control'] !== 'no-store' ||
    count('content-length') !== 1 ||
    count('transfer-encoding') ||
    count('content-encoding') ||
    typeof headers['content-length'] !== 'string' ||
    !/^[0-9]+$/.test(headers['content-length'])
  )
    throw new ContainerCredentialError('invalid_response');
  const length = Number(headers['content-length']);
  if (!Number.isInteger(length) || length <= 0 || length > MAX_BODY) {
    throw new ContainerCredentialError('invalid_response');
  }
  return length;
}

/** Own the request and its socket through terminal close, on every outcome. */
async function readToken(
  endpoint: Endpoint,
  target: string,
  signal?: AbortSignal
): Promise<Buffer> {
  if (signal?.aborted) throw new ContainerCredentialError('aborted');
  return new Promise((resolve, reject) => {
    const body = JSON.stringify({ target_database: target });
    let req: ClientRequest | undefined;
    let socket: Socket | undefined;
    let requestClosed = false;
    let socketClosed = false;
    let outcome:
      | { error?: ContainerCredentialError; bytes?: Buffer }
      | undefined;
    let timer: ReturnType<typeof setTimeout> | undefined;
    const complete = (): void => {
      if (!outcome || !requestClosed || (socket && !socketClosed)) return;
      clearTimeout(timer);
      signal?.removeEventListener('abort', abort);
      if (outcome.error) reject(outcome.error);
      else resolve(outcome.bytes!);
    };
    const finish = (error?: ContainerCredentialError, bytes?: Buffer): void => {
      if (!outcome) outcome = { error, bytes };
      req?.destroy();
      socket?.destroy();
      complete();
    };
    const fail = (code: ContainerCredentialErrorCode): void =>
      finish(new ContainerCredentialError(code));
    const abort = (): void => fail('aborted');
    try {
      req = request({
        ...endpoint,
        // A fresh private agent owns one connection; ambient proxy and HTTP
        // pool settings cannot redirect this credential request.
        agent: false,
        method: 'POST',
        path: CREDENTIAL_PATH,
        maxHeaderSize: MAX_HEADERS,
        headers: {
          Host: 'socketPath' in endpoint ? LOCAL_AUTHORITY : endpoint.host,
          'Content-Type': 'application/json',
          'Content-Length': Buffer.byteLength(body),
          Connection: 'close',
        },
      });
      req.once('socket', owned => {
        socket = owned;
        owned.once('close', () => {
          socketClosed = true;
          complete();
        });
        if (outcome) owned.destroy();
      });
      req.once('error', () => fail('transport'));
      req.once('close', () => {
        requestClosed = true;
        if (!outcome) fail('transport');
        complete();
      });
      req.once('response', response => {
        response.on('error', () => fail('transport'));
        let length: number;
        try {
          length = responseLength(response);
        } catch (error) {
          finish(
            error instanceof ContainerCredentialError
              ? error
              : new ContainerCredentialError('invalid_response')
          );
          return;
        }
        const chunks: Buffer[] = [];
        let received = 0;
        response.on('data', (chunk: Buffer) => {
          received += chunk.length;
          if (received > length) {
            fail('invalid_response');
            return;
          }
          chunks.push(chunk);
        });
        response.once('aborted', () => fail('transport'));
        response.once('end', () => {
          if (received !== length) fail('invalid_response');
          else finish(undefined, Buffer.concat(chunks, received));
        });
      });
      // Upgrades never become an untracked or credential-bearing connection.
      req.once('upgrade', (_response, upgraded) => {
        upgraded.destroy();
        fail('invalid_response');
      });
      timer = setTimeout(() => fail('timeout'), REQUEST_MS);
      signal?.addEventListener('abort', abort, { once: true });
      if (signal?.aborted) abort();
      else req.end(body);
    } catch {
      if (req) fail('transport');
      else {
        clearTimeout(timer);
        signal?.removeEventListener('abort', abort);
        reject(new ContainerCredentialError('transport'));
      }
    }
  });
}

File: sdks/rust/src/credentials/container.rs
//! Explicit platform discovery and bounded, target-specific credential requests.

use crate::Identity;
use http::Uri;
use reqwest::{header, redirect::Policy, Client, Response, Url};
use serde::Deserialize;
use std::{
    fmt,
    net::IpAddr,
    path::{Component, PathBuf},
    time::{Duration, SystemTime, UNIX_EPOCH},
};
use tokio::time::Instant;

const DATABASE_IDENTITY: &str = "SPACETIMEDB_DATABASE_IDENTITY";
const SERVER_URI: &str = "SPACETIMEDB_SERVER_URI";
const CREDENTIAL_BROKER: &str = "SPACETIMEDB_CREDENTIAL_BROKER";
const REQUEST_TIMEOUT: Duration = Duration::from_secs(5);
const MAX_LIFETIME: Duration = Duration::from_secs(30);
const MAX_TOKEN: usize = 8192;
const MAX_BODY: usize = MAX_TOKEN + 256;
const MAX_HEADERS: usize = 4096;
const MAX_URI: usize = 4096;
// The Unix transport uses the same HTTP authority and request path as the
// guest proxy. This URL is never resolved or connected over TCP in Unix mode.
const LOCAL_HTTP_URI: &str = "http://127.0.0.1:18081/v1/credentials";

/// A credential error whose diagnostics never include discovery values,
/// response bodies, request headers, or tokens.
#[derive(Clone, Copy, Debug, thiserror::Error, PartialEq, Eq)]
#[non_exhaustive]
pub enum ContainerCredentialError {
    #[error("Missing or non-Unicode container discovery variable {0}")]
    MissingEnvironment(&'static str),
    #[error("Invalid container discovery configuration")]
    InvalidDiscovery,
    #[error("Unsupported container credential transport")]
    UnsupportedTransport,
    #[error("Container credentials cannot be combined with a static token")]
    ConflictingCredentials,
    #[error("Container credentials cannot be used with debug files that record tokens")]
    DebugLogging,
    #[error("Invalid container credential target")]
    InvalidTarget,
    #[error("Unable to resolve the database Identity on the selected server")]
    NameResolution,
    #[error("Container credential request was denied")]
    Denied,
    #[error("Container credential broker is unavailable")]
    Unavailable,
    #[error("Container credential request failed")]
    Transport,
    #[error("Container credential request exceeded its deadline")]
    Timeout,
    #[error("Invalid container credential response")]
    InvalidResponse,
    #[error("Container credential expired before the connection completed")]
    Expired,
}

type Result<T> = std::result::Result<T, ContainerCredentialError>;

#[derive(Clone)]
enum Endpoint {
    Http(Url),
    Unix(PathBuf),
}

/// Discovery configuration for a process running in a SpacetimeDB container.
///
/// This value contains no credentials. Cloning it never copies a token. The
/// broker derives the sender from its authenticated runtime association, not
/// from the Identity in this configuration or from request fields.
///
/// ```ignore
/// use spacetimedb_sdk::credentials;
/// let container = credentials::Container::from_env()?;
/// let connection = DbConnection::builder()
///     .with_container_credentials(container)
///     .build_async().await?;
/// ```
///
/// The builder defaults to this container's database and server. Use its
/// `with_uri` and `with_database_name` methods to select another database.
/// Explicit `self` resolves locally to [`Self::database_identity`].
///
/// Each builder connection requests a fresh token. To renew credentials and
/// reconnect with fresh subscriptions, use [`crate::ContainerSession`]. An
/// ordinary builder connection ends when its credential expires or is revoked.
/// Do not save the token returned by `on_connect` for later use.
#[derive(Clone)]
pub struct Container {
    identity: Identity,
    server_uri: Uri,
    endpoint: Endpoint,
}

impl fmt::Debug for Container {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Container").finish_non_exhaustive()
    }
}

impl Container {
    /// Read only the three platform-injected discovery variables. Missing or
    /// invalid configuration is an error; this never reads stored CLI/SDK
    /// credentials, allocates an anonymous Identity, or chooses a default server.
    /// No network or socket operation occurs until a token is requested.
    pub fn from_env() -> Result<Self> {
        let read = |key| std::env::var(key).map_err(|_| ContainerCredentialError::MissingEnvironment(key));
        Self::new(&read(DATABASE_IDENTITY)?, &read(SERVER_URI)?, &read(CREDENTIAL_BROKER)?)
    }

    /// Use explicit discovery values instead of reading the environment.
    ///
    /// The broker must be an absolute `unix:///path` socket URI or an HTTP URI
    /// with a numeric loopback address and the path `/v1/credentials`. Unix
    /// sockets are supported on Unix targets. Broker redirects, proxies, DNS
    /// lookup, embedded credentials, and fallback transports are forbidden.
    pub fn new(database_identity: &str, server_uri: &str, broker_uri: &str) -> Result<Self> {
        let identity = parse_identity(database_identity).map_err(|_| ContainerCredentialError::InvalidDiscovery)?;
        let server_uri = server_url(server_uri)?
            .as_str()
            .parse()
            .map_err(|_| ContainerCredentialError::InvalidDiscovery)?;
        Ok(Self {
            identity,
            server_uri,
            endpoint: endpoint(broker_uri)?,
        })
    }

    pub fn database_identity(&self) -> Identity {
        self.identity
    }

    pub fn server_uri(&self) -> &Uri {
        &self.server_uri
    }

    /// Obtain a fresh credential for HTTP or another client of `target`.
    ///
    /// Send it only to that database's trusted server. It is a bearer credential
    /// with a lifetime of at most 30 seconds, possibly less. Keep it in memory;
    /// never put it in environment variables, logs, images, or durable files.
    /// This method neither retries a denial nor falls back to an owner token.
    pub async fn token_for(&self, target: Identity) -> Result<ContainerToken> {
        tokio::time::timeout(REQUEST_TIMEOUT, self.request_token(target))
            .await
            .map_err(|_| ContainerCredentialError::Timeout)?
    }

    async fn request_token(&self, target: Identity) -> Result<ContainerToken> {
        let builder = client_builder();
        let (builder, uri) = match &self.endpoint {
            Endpoint::Http(uri) => (builder, uri.clone()),
            Endpoint::Unix(path) => {
                #[cfg(unix)]
                {
                    (builder.unix_socket(path.clone()), Url::parse(LOCAL_HTTP_URI).unwrap())
                }
                #[cfg(not(unix))]
                {
                    let _ = path;
                    return Err(ContainerCredentialError::UnsupportedTransport);
                }
            }
        };
        let client = builder.build().map_err(|_| ContainerCredentialError::Transport)?;
        // Identity::to_hex is fixed lowercase ASCII, so the body cannot contain
        // request framing, another sender, or arbitrary JSON fields.
        let body = format!("{{\"target_database\":\"{}\"}}", target.to_hex());
        let response = client
            .post(uri)
            .header(header::CONTENT_TYPE, "application/json")
            .header(header::CONNECTION, "close")
            .body(body)
            .send()
            .await
            .map_err(transport_error)?;
        match response.status().as_u16() {
            200 => {}
            401 | 403 => return Err(ContainerCredentialError::Denied),
            503 => return Err(ContainerCredentialError::Unavailable),
            _ => return Err(ContainerCredentialError::InvalidResponse),
        }
        let headers = response.headers();
        if header_size(headers) > MAX_HEADERS
            || headers.get_all(header::CONTENT_TYPE).iter().count() != 1
            || headers.get(header::CONTENT_TYPE).map(|value| value.as_bytes()) != Some(b"application/json")
            || headers.get_all(header::CACHE_CONTROL).iter().count() != 1
            || headers.get(header::CACHE_CONTROL).map(|value| value.as_bytes()) != Some(b"no-store")
            || headers.contains_key(header::TRANSFER_ENCODING)
            || headers.contains_key(header::CONTENT_ENCODING)
            || headers.get_all(header::CONTENT_LENGTH).iter().count() != 1
        {
            return Err(ContainerCredentialError::InvalidResponse);
        }
        let length = headers[header::CONTENT_LENGTH]
            .to_str()
            .ok()
            .filter(|value| !value.is_empty() && value.bytes().all(|byte| byte.is_ascii_digit()))
            .and_then(|value| value.parse::<usize>().ok())
            .filter(|length| *length > 0 && *length <= MAX_BODY)
            .ok_or(ContainerCredentialError::InvalidResponse)?;
        let bytes = bounded_body(response, length).await?;
        if bytes.len() != length {
            return Err(ContainerCredentialError::InvalidResponse);
        }
        let response: TokenResponse =
            serde_json::from_slice(&bytes).map_err(|_| ContainerCredentialError::InvalidResponse)?;
        let expiry = UNIX_EPOCH
            .checked_add(Duration::from_secs(response.expires_unix_seconds))
            .ok_or(ContainerCredentialError::InvalidResponse)?;
        let remaining = expiry
            .duration_since(SystemTime::now())
            .ok()
            .filter(|remaining| !remaining.is_zero() && *remaining <= MAX_LIFETIME)
            .ok_or(ContainerCredentialError::InvalidResponse)?;
        if response.token.is_empty()
            || response.token.len() > MAX_TOKEN
            || !response.token.bytes().all(|byte| byte.is_ascii_graphic())
        {
            return Err(ContainerCredentialError::InvalidResponse);
        }
        // Retain a monotonic cap even if the system clock later moves backward.
        let deadline = Instant::now() + remaining;
        Ok(ContainerToken {
            token: response.token,
            target,
            expiry,
            deadline,
        })
    }

    pub(crate) async fn prepare_connection(&self, uri: Option<&Uri>, target: Option<&str>) -> Result<Prepared> {
        let (uri, target) = self.resolve_connection_target(uri, target).await?;
        let credential = self.token_for(target).await?;
        Ok(Prepared {
            uri,
            target: target.to_hex().to_string(),
            credential,
        })
    }

    pub(crate) async fn resolve_connection_target(
        &self,
        uri: Option<&Uri>,
        target: Option<&str>,
    ) -> Result<(Uri, Identity)> {
        let uri = uri.unwrap_or(&self.server_uri);
        let mut server = server_url(&uri.to_string())?;
        let target = match target {
            None | Some("self") => self.identity,
            Some(target) => match parse_identity(target) {
                Ok(identity) => identity,
                Err(_) => {
                    // Preserve the SDK's server path prefix. Append the name as
                    // one encoded segment, never as request path syntax.
                    if target.is_empty() || target.len() > 256 || target.chars().any(char::is_control) {
                        return Err(ContainerCredentialError::InvalidTarget);
                    }
                    server
                        .path_segments_mut()
                        .map_err(|_| ContainerCredentialError::InvalidDiscovery)?
                        .pop_if_empty()
                        .extend(["v1", "database", target, "identity"]);
                    resolve_name(server).await?
                }
            },
        };
        Ok((uri.clone(), target))
    }
}

pub(crate) struct Prepared {
    pub uri: Uri,
    pub target: String,
    pub credential: ContainerToken,
}

/// An in-memory, target-specific bearer credential. Debug output is redacted.
/// Expiry is enforced by the receiving host even if its bytes are copied.
pub struct ContainerToken {
    token: String,
    target: Identity,
    expiry: SystemTime,
    deadline: Instant,
}

impl ContainerToken {
    /// The token for an `Authorization: Bearer ...` header. Do not log or save it.
    pub fn as_str(&self) -> &str {
        &self.token
    }

    pub fn target(&self) -> Identity {
        self.target
    }

    pub fn expires_at(&self) -> SystemTime {
        self.expiry
    }

    pub fn remaining_lifetime(&self) -> Duration {
        self.deadline.saturating_duration_since(Instant::now())
    }

    pub(crate) fn deadline(&self) -> Instant {
        self.deadline
    }
}

impl fmt::Debug for ContainerToken {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("ContainerToken")
            .field("target", &self.target)
            .field("expires_at", &self.expiry)
            .field("token", &"<redacted>")
            .finish()
    }
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct TokenResponse {
    token: String,
    expires_unix_seconds: u64,
}

fn client_builder() -> reqwest::ClientBuilder {
    Client::builder()
        .no_proxy()
        .redirect(Policy::none())
        .referer(false)
        .http1_only()
        .pool_max_idle_per_host(0)
        .connect_timeout(REQUEST_TIMEOUT)
        .timeout(REQUEST_TIMEOUT)
        .no_gzip()
        .no_brotli()
        .no_deflate()
        .no_zstd()
}

fn transport_error(error: reqwest::Error) -> ContainerCredentialError {
    if error.is_timeout() {
        ContainerCredentialError::Timeout
    } else {
        ContainerCredentialError::Transport
    }
}

async fn bounded_body(mut response: Response, limit: usize) -> Result<Vec<u8>> {
    let mut bytes = Vec::new();
    while let Some(chunk) = response.chunk().await.map_err(transport_error)? {
        if chunk.len() > limit.saturating_sub(bytes.len()) {
            return Err(ContainerCredentialError::InvalidResponse);
        }
        bytes.extend_from_slice(&chunk);
    }
    Ok(bytes)
}

fn header_size(headers: &header::HeaderMap) -> usize {
    headers
        .iter()
        .map(|(key, value)| key.as_str().len() + value.as_bytes().len() + 4)
        .sum()
}

async fn resolve_name(uri: Url) -> Result<Identity> {
    tokio::time::timeout(REQUEST_TIMEOUT, async {
        let client = client_builder()
            .build()
            .map_err(|_| ContainerCredentialError::NameResolution)?;
        let response = client
            .get(uri)
            .send()
            .await
            .map_err(|_| ContainerCredentialError::NameResolution)?;
        if response.status() != reqwest::StatusCode::OK || header_size(response.headers()) > MAX_HEADERS {
            return Err(ContainerCredentialError::NameResolution);
        }
        let body = bounded_body(response, 64)
            .await
            .map_err(|_| ContainerCredentialError::NameResolution)?;
        parse_identity(std::str::from_utf8(&body).map_err(|_| ContainerCredentialError::NameResolution)?)
            .map_err(|_| ContainerCredentialError::NameResolution)
    })
    .await
    .map_err(|_| ContainerCredentialError::Timeout)?
}

fn parse_identity(value: &str) -> Result<Identity> {
    if value.len() != 64 || !value.bytes().all(|byte| byte.is_ascii_hexdigit()) {
        return Err(ContainerCredentialError::InvalidTarget);
    }
    Identity::from_hex(value).map_err(|_| ContainerCredentialError::InvalidTarget)
}

fn parse_url(value: &str) -> Result<Url> {
    if value.is_empty() || value.len() > MAX_URI || value.bytes().any(|byte| byte.is_ascii_control() || byte == b' ') {
        return Err(ContainerCredentialError::InvalidDiscovery);
    }
    let uri = Url::parse(value).map_err(|_| ContainerCredentialError::InvalidDiscovery)?;
    if !uri.username().is_empty() || uri.password().is_some() || uri.query().is_some() || uri.fragment().is_some() {
        return Err(ContainerCredentialError::InvalidDiscovery);
    }
    Ok(uri)
}

fn server_url(value: &str) -> Result<Url> {
    let mut uri = parse_url(value)?;
    match uri.scheme() {
        "http" | "https" => {}
        "ws" => uri
            .set_scheme("http")
            .map_err(|_| ContainerCredentialError::InvalidDiscovery)?,
        "wss" => uri
            .set_scheme("https")
            .map_err(|_| ContainerCredentialError::InvalidDiscovery)?,
        _ => return Err(ContainerCredentialError::InvalidDiscovery),
    }
    if uri.host_str().is_none() {
        return Err(ContainerCredentialError::InvalidDiscovery);
    }
    Ok(uri)
}

fn endpoint(value: &str) -> Result<Endpoint> {
    let uri = parse_url(value)?;
    match uri.scheme() {
        "http" => {
            // Parse the original authority as an IP, rejecting DNS aliases and
            // URL parser shortcuts such as hexadecimal/abbreviated IPv4.
            let original: Uri = value.parse().map_err(|_| ContainerCredentialError::InvalidDiscovery)?;
            let host = original.host().ok_or(ContainerCredentialError::InvalidDiscovery)?;
            let ip: IpAddr = host
                .trim_matches(['[', ']'])
                .parse()
                .map_err(|_| ContainerCredentialError::InvalidDiscovery)?;
            if !ip.is_loopback() || uri.path() != "/v1/credentials" {
                return Err(ContainerCredentialError::InvalidDiscovery);
            }
            Ok(Endpoint::Http(uri))
        }
        "unix" => {
            let path = value
                .strip_prefix("unix://")
                .ok_or(ContainerCredentialError::InvalidDiscovery)?;
            if uri.host_str().is_some()
                || uri.port().is_some()
                || !path.starts_with('/')
                || path.len() > 103
                || path.contains('%')
                || path.split('/').any(|part| part == "." || part == "..")
                || PathBuf::from(path)
                    .components()
                    .any(|part| !matches!(part, Component::RootDir | Component::Normal(_)))
            {
                return Err(ContainerCredentialError::InvalidDiscovery);
            }
            #[cfg(unix)]
            return Ok(Endpoint::Unix(PathBuf::from(path)));
            #[cfg(not(unix))]
            return Err(ContainerCredentialError::UnsupportedTransport);
        }
        _ => Err(ContainerCredentialError::UnsupportedTransport),
    }
}

#[cfg(test)]
pub(crate) mod tests;
